### ISSUE TITLE
Equivalent switch expression produces suboptimal IL compared to switch statement

### ISSUE THREAD

**Issue by bobbymcr** [dotnet/roslyn#44546]
**Version Used**: 
.NET Core 3.1.3 (CoreCLR 4.700.20.11803, CoreFX 4.700.20.12001), X64 RyuJIT

This code uses a switch statement to convert an enum value to a char:

    public static class MyLevel
    {
        public static char Translate1(TraceLevel level)
        {
            switch (level)
            {
                case TraceLevel.Verbose: return 'V';
                case TraceLevel.Info: return 'I';
                case TraceLevel.Warning: return 'W';
                case TraceLevel.Error: return 'E';
                case TraceLevel.Off: return ' ';
                default: throw new ArgumentOutOfRangeException(nameof(level));
            };
        }
    }

In Visual Studio 2019, this is flagged by IDE066 ("Use 'switch' expression") and offers the code fix "Convert switch statement to expression". Applying the code fix results in this translation:

    public static class MyLevel
    {
        public static char Translate2(TraceLevel level)
        {
            return level switch
            {
                TraceLevel.Verbose => 'V',
                TraceLevel.Info => 'I',
                TraceLevel.Warning => 'W',
                TraceLevel.Error => 'E',
                TraceLevel.Off => ' ',
                _ => throw new ArgumentOutOfRangeException(nameof(level)),
            };
        }
    }

However, this code performs slightly worse than the original switch statement version when compared in a micro-benchmark:

    [SimpleJob(RuntimeMoniker.NetCoreApp31)]
    [MemoryDiagnoser]
    public class SwitchBenchmark
    {
        [Benchmark]
        public int Switch1()
        {
            int sum = 0;
            sum += MyLevel.Translate1(TraceLevel.Error);
            sum += MyLevel.Translate1(TraceLevel.Warning);
            sum += MyLevel.Translate1(TraceLevel.Info);
            sum += MyLevel.Translate1(TraceLevel.Verbose);
            sum += MyLevel.Translate1(TraceLevel.Off);
            return sum;
        }

        [Benchmark]
        public int Switch2()
        {
            int sum = 0;
            sum += MyLevel.Translate2(TraceLevel.Error);
            sum += MyLevel.Translate2(TraceLevel.Warning);
            sum += MyLevel.Translate2(TraceLevel.Info);
            sum += MyLevel.Translate2(TraceLevel.Verbose);
            sum += MyLevel.Translate2(TraceLevel.Off);
            return sum;
        }
    }

Benchmark results:

    BenchmarkDotNet=v0.12.1, OS=Windows 10.0.18363.836 (1909/November2018Update/19H2)
    Intel Core i7-3960X CPU 3.30GHz (Ivy Bridge), 1 CPU, 12 logical and 6 physical cores
    .NET Core SDK=3.1.201
      [Host]        : .NET Core 3.1.3 (CoreCLR 4.700.20.11803, CoreFX 4.700.20.12001), X64 RyuJIT
      .NET Core 3.1 : .NET Core 3.1.3 (CoreCLR 4.700.20.11803, CoreFX 4.700.20.12001), X64 RyuJIT

|  Method |     Mean |    Error |   StdDev |   Median | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------- |---------:|---------:|---------:|---------:|------:|------:|------:|----------:|
| Switch1 | 15.21 ns | 0.368 ns | 1.008 ns | 14.89 ns |     - |     - |     - |         - |
| Switch2 | 16.22 ns | 0.253 ns | 0.211 ns | 16.14 ns |     - |     - |     - |         - |

The reason for this seems to be that the IL code produced by the switch expression fails to take advantage of early returns and does a local store with branch instead:

    .method public hidebysig static 
        char Translate1 (
            valuetype [System.Diagnostics.TraceSource]System.Diagnostics.TraceLevel level
        ) cil managed 
    {
        // Method begins at RVA 0x2113
        // Code size 54 (0x36)
        .maxstack 8

        // switch (level)
        IL_0000: ldarg.0
        // (no C# code)
        IL_0001: switch (IL_0028, IL_0025, IL_0022, IL_001f, IL_001c)

        IL_001a: br.s IL_002b

        // return 'V';
        IL_001c: ldc.i4.s 86
        // (no C# code)
        IL_001e: ret

        // return 'I';
        IL_001f: ldc.i4.s 73
        // (no C# code)
        IL_0021: ret

        // return 'W';
        IL_0022: ldc.i4.s 87
        // (no C# code)
        IL_0024: ret

        // return 'E';
        IL_0025: ldc.i4.s 69
        // (no C# code)
        IL_0027: ret

        // return ' ';
        IL_0028: ldc.i4.s 32
        // (no C# code)
        IL_002a: ret

        // throw new ArgumentOutOfRangeException("level");
        IL_002b: ldstr "level"
        IL_0030: newobj instance void [System.Runtime]System.ArgumentOutOfRangeException::.ctor(string)
        // (no C# code)
        IL_0035: throw
    } // end of method MyLevel::Translate1

    .method public hidebysig static 
        char Translate2 (
            valuetype [System.Diagnostics.TraceSource]System.Diagnostics.TraceLevel level
        ) cil managed 
    {
        // Method begins at RVA 0x214c
        // Code size 66 (0x42)
        .maxstack 1
        .locals init (
            [0] char
        )

        // switch (level)
        IL_0000: ldarg.0
        // (no C# code)
        IL_0001: switch (IL_0030, IL_002b, IL_0026, IL_0021, IL_001c)

        IL_001a: br.s IL_0035

        // return 'V';
        IL_001c: ldc.i4.s 86
        IL_001e: stloc.0
        // (no C# code)
        IL_001f: br.s IL_0040

        // return 'I';
        IL_0021: ldc.i4.s 73
        IL_0023: stloc.0
        // (no C# code)
        IL_0024: br.s IL_0040

        // return 'W';
        IL_0026: ldc.i4.s 87
        IL_0028: stloc.0
        // (no C# code)
        IL_0029: br.s IL_0040

        // return 'E';
        IL_002b: ldc.i4.s 69
        IL_002d: stloc.0
        // (no C# code)
        IL_002e: br.s IL_0040

        // return ' ';
        IL_0030: ldc.i4.s 32
        IL_0032: stloc.0
        // (no C# code)
        IL_0033: br.s IL_0040

        // throw new ArgumentOutOfRangeException("level");
        IL_0035: ldstr "level"
        IL_003a: newobj instance void [System.Runtime]System.ArgumentOutOfRangeException::.ctor(string)
        // (no C# code)
        IL_003f: throw

        IL_0040: ldloc.0
        IL_0041: ret
    } // end of method MyLevel::Translate2

Is it possible to optimize this pattern so that it generates the same type of IL code for both the switch statement and switch expression?

**Comment by Sewer56** [comment]
Ran across this while working on a project, here's a [SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEBmAAgFcA7AZwEMAzGUwgJlIGFSBvAWACh+pIaQD0Y0gB5q5ALYzaUAJ4A+QcLEjSAFUUAHJhHqkAJrQy1S1DNBjHS2SqQwALbNVL1sMXMYB0aoQ1JEWk5BRUA5jIYSlkOCFxZSh19UhBSYEUMGEjePmEC0gAhLKYAXlIABjRIgoBlEuzSCoBGGvzCoQBVAElKDBakZtIWds6hPoGhiuIx8d7+4jYK9FrhSaXhgFY5zoWMVGGkXcLJw4qAdhP6hwBzXHLSAA5r4QARCHJgB+GATlehHUMFA7sMWtU1kIAEq0ADubzMFlabUhpAA4uRsHZWqNIgBfSKRIJSWTyJSqDqBcRomAYdwuJgPSi3FykQxOZxMAButESTAcGVK7gAFNgjLpYNQYhgAJT+SmicQSEKk8IUgokCg0Bj8/ro2kANV55BgABkYiznML2AkkikmAB9DB6GAy3KowgXUhOl2WWHYDBgZyovLjQo2xIyZIu3yNcrKUgohVhiN2mMNUrNBNJsMFVNR+2+faDLMjAHh20FmOTEtlBO45PjfPR/RFjbLBOrRudZuF9ulru54S9mP7c4Jl6ovOVlswXxnFClyfdiuR2e+Op3H510iDofxNeFj5fbcT8vTw/p4Ggnd7ocj1sw+GIpfn4czwsYrGlwZvoQO0tKlRPEAG58UJBViVCMkIkg6laXpTlSCZS02SMBlSB5Pl7EcTJshFMVSAlGApX6OUiSVFUwnJSJNSoOhGBwjB9QwI0+XNZkXAAMVoKxrQ/X0fX0N0FVDcZqH9QNnFIYUhNdEMp2EMBeKYB85zjdJPUTMCVwKZSpQPNNWwzbJNK9FodP3Uh9NUgTW2LJAzJGSz9xswyq1bGtHOYL0WBcoc3LUttFhYJyUH83NArsud2zCiKwyiy97IXJynni8ZEqMmL+lQVL0s6TKPLnTdmQeOLFKEQr12Pb4YDyirrJU9z1yBEFmXK3SlKaoKnwRcx6s6yruui3wv2MJzBnywpjBgehaHIXAsB8qopoJBU1rxIA=) to reproduce.

In case SharpLab ceases to exist one day:

```csharp
public unsafe class C {

    /// <summary>
    /// Type of data stored in this field.
    /// </summary>
    public enum ColumnType : byte
    {
        Byte = 0,
        SByte = 1,
        UInt16 = 2,
        Int16 = 3,
        UInt32 = 4,
        Int32 = 5,
        UInt64 = 6,
        Int64 = 7,
        Single = 8,
        Double = 9,
        String = 10,
        RawData = 11,
        Guid = 12,
    }
    
    /// <summary>
    /// Gets the length of the value in bytes (if present).
    /// </summary>
    public unsafe int GetValueLength(ColumnType _type)
    {
        return _type switch
        {
            ColumnType.Byte => 1,
            ColumnType.SByte => 1,
            ColumnType.UInt16 => 2,
            ColumnType.Int16 => 2,
            ColumnType.UInt32 => 4,
            ColumnType.Int32 => 4,
            ColumnType.UInt64 => 8,
            ColumnType.Int64 => 8,
            ColumnType.Single => 4,
            ColumnType.Double => 8,
            ColumnType.String => 4,
            ColumnType.RawData => 8,
            ColumnType.Guid => 16,
            _ => 0
        };
    }
    
    /// <summary>
    /// Gets the length of the value in bytes (if present).
    /// </summary>
    public unsafe int GetValueLengthFast(ColumnType _type)
    {
        switch (_type)
        {
            case ColumnType.Byte: return 1;
            case ColumnType.SByte: return 1;
            case ColumnType.UInt16: return 2;
            case ColumnType.Int16: return 2;
            case ColumnType.UInt32: return 4;
            case ColumnType.Int32: return 4;
            case ColumnType.UInt64: return 8;
            case ColumnType.Int64: return 8;
            case ColumnType.Single: return 4;
            case ColumnType.Double: return 8;
            case ColumnType.String: return 4;
            case ColumnType.RawData: return 8;
            case ColumnType.Guid: return 16;
            default: return 0;
        }
    }
}
```

Unoptimal IL pattern is same as original post.
